mimxrt: A set of small extensions and a bug fix.

robert-hh opened the pull request:
This is another service type pull request. The commits cover:
- support gaps in the I2C and SPI device numbers, allowing e.g. to define SPI(1) and SPI(2) instead of SPI(0) and SPI(1). Then, the SPI (I2C) numbers can match a boards Pin-out document.
- Allow for board-specific flash config files instead of the common ones.
- Compensate for a bug in the NXP lib SPI driver, which fails when short read/write commands alternate.
- Remove two obsolete files from the Seeed Arch Mix file set.
- Replace numeric constants for the Pin configuration by a function call, preparing for MCU's with different Pin config bit definitions, like the MIMXRT 117x series.
- Allow to specify the CS signal used for SPI, using the keyword options cs=nn. The default is 0, the only other value implemented is 1. Only a few boards have the CS1 pins exposed to the board connector. Required e.g. for the upcoming Olimex board.
- Allow configuring USB ID Elements Manufacturer, Board name, Unique-ID, Vendor ID and Product ID. Replaces #8094.
- Adding support for the Olimex RT1010 board
- Simplify Makefile a little bit.

robert-hh committed:
mimxrt: Support gaps in the SPI an I2C device numbers.

That allows also to use e.g. SPI1 and SPI2 instead SPI0 and SPI1.

robert-hh committed:
mimxrt: Allow for board-specific flash driver files.

If in a board's mkconfigboard.mk the following symbol is set:

MICROPY_HW_BOARD_FLASH_FILES = 1

then the files:

($BOARD)_flexspi_flash_config.h  and
qspi_nor_flash_config.c  and/or
qspi_hyper_flash_config.c

are expected in the board directory. Otherwise the common files from
the hal directory are used.

robert-hh committed:
mimxrt: Compensate for a bug in the fsl_lpspi.c file.

This library file has a bug, in that TransferBlocking returns before the
transfer has finished. That is a problem if a write follows immediately
a read.

robert-hh committed:
mimxrt: Remove two files from the Seeed Arch Mix directory.

These were leftovers from a previous refactoring in the Seeed Arch Mix
directory.

robert-hh committed:
mimxrt: Replace Pin-config constants by a function call.

The Pin config setting by IOMUXC_SetPinConfig() is supplied by a
bit pattern. That pattern is specific for a MCU family. In preparation
for supporting the MIMXRT117x family, the constant bit pattern is
replaced by a function call, such that the bit pattern is created
at a single place. The code for this functions was taken from
machine_pin.c.

Note: A working port for the MIMXRT1176 exists already.

robert-hh committed:
mimxrt: Allow to select cs0 or cs1 for SPI.

Using the keyword argument cs=nnn in the constructor. The cs1
pin has to be defined in mpconfigboard.h.
Note: Only a few boards have the CS1 pin exposed to the connectors.

robert-hh committed:
mimxrt: Add USB ID elements.

- Manufacturer, set by MICROPY_HW_USB_STR_MANUF; default "MicroPyhton"
- Board name, as set by MICROPY_HW_BOARD_NAME
- Unique-ID, same as returned by machine.unique_id().
- USB Vendor ID, as set by MICROPY_HW_USB_VID; default0xf055
- USB Product ID, as set by MICROPY_HW_USB_PID; default 0x9802

robert-hh committed:
mimxrt: Add support for the Olimex RT1010 board.

The board.json file is intentionally excluded, until the board will be
sold. But including it into the mimxrt series make it easier to keep
the build up-to-date.

alphaFred reviewed (commented):
in `ports/mimxrt/Makefile` I would suggest that we move the selection of flash type and flash configuration together into one block like so:
```
ifeq ($(MICROPY_HW_FLASH_TYPE), qspi_nor)
CFLAGS += -DBOARD_FLASH_OPS_HEADER_H=\"hal/flexspi_nor_flash.h\"
SRC_C += hal/flexspi_nor_flash.c

ifeq ($(MICROPY_HW_BOARD_FLASH_FILES), 1)
CFLAGS += -DBOARD_FLASH_CONFIG_HEADER_H=\"$(BOARD)_flexspi_flash_config.h\"
SRC_C += $(BOARD_DIR)/qspi_nor_flash_config.c
else
CFLAGS += -DBOARD_FLASH_CONFIG_HEADER_H=\"hal/flexspi_flash_config.h\"
SRC_C += hal/qspi_nor_flash_config.c
endif
#
else ifeq ($(MICROPY_HW_FLASH_TYPE), hyperflash)
CFLAGS += -DBOARD_FLASH_OPS_HEADER_H=\"hal/flexspi_hyper_flash.h\"
SRC_C += hal/flexspi_hyper_flash.c

ifeq ($(MICROPY_HW_BOARD_FLASH_FILES), 1)
CFLAGS += -DBOARD_FLASH_CONFIG_HEADER_H=\"$(BOARD)_flexspi_flash_config.h\"
SRC_C += $(BOARD_DIR)/qspi_hyper_flash_config.c
else
CFLAGS += -DBOARD_FLASH_CONFIG_HEADER_H=\"hal/flexspi_flash_config.h\"
SRC_C += hal/qspi_hyper_flash_config.c
endif
#
else
$(error Error: Unknown board flash type $(MICROPY_HW_FLASH_TYPE))
endif
```

robert-hh commented:
Looks OK and works.

robert-hh committed:
mimxrt: Simplify the Makefile.

Simplify it a little bit by combining two sections regarding the
flash type. Thanks to @alphaFred for suggesting it.

dpgeorge commented:
Thanks for this, looks good, and nice and clean commits.  Rebased and merged in 2dd3d88409937908da6ee05d2d391453df9a3c8a through c1841c2d455a99f4184487da658e88267785ebd5

I also tested on Teensy 4.0, the USB serial number/name works well!
